### PR TITLE
Implement threading to speed up rose-ana (PR on Master)

### PR DESCRIPTION
**Pull request by stevewardle**
We have a few cases in our test suite of rose-ana taking a long time - typically when doing lots of file comparisons (in the worst case a rose-ana task with ~20 operations had 4 or 5 that took ~10 minutes each, for a total runtime of over an hour)

It seems like an easy win for all rose-ana users to make it able to run operations in parallel.  Given the way the analysis tasks are written (as classes/plugins) it seems it would lend itself most nicely to multi-threading rather than multi-process (despite the process approach generally being easier to manage)

### CODE REVIEW

File: lib/python/rose/apps/rose_ana.py [hunk 1]
```diff
@@ -293,58 +381,114 @@ def run(self, app_runner, conf_tree, opts, args, uuid, work_files):
...
+                number_of_failures += thread.failures
+                number_of_skips += thread.skips
+                task_error = (task_error or thread.task_error)
+                summary_status += thread.summary_status
+                # And print the output via the main reporter
+                for args in thread.reporter_args:
+                    self.reporter(*args)
+
+        else:
+            # Negative threads?
```
Comment (wxtim) [inline lib/python/rose/apps/rose_ana.py]: Would
```python
msg = (f"Number of threads must be a positive integer, "
       f"so should not be {n_threads}")
```
cover more cases, and offer more assistance to the end user? 

Could you add tests for n_threads = [-1, 0, 0.25, "Tim"]) to check this loop?

**Comment by matthewrmshin** [comment]
Hi @stevewardle please do another re-based of this branch. Travis CI's issue should be fixed on latest 2019.01.x branch.

**Comment by stevewardle** [comment]
> Hi @stevewardle please do another re-based of this branch. Travis CI's issue should be fixed on latest 2019.01.x branch.

Hasn't it been merged? (see above)

**Comment by matthewrmshin** [comment]
Sorry I meant re-base against master, please?

**Comment by oliver-sanders** [comment]
Note that the `rose_ana` tests won't run until #2317 is merged.

**Comment by matthewrmshin** [comment]
Branch in conflict.

**Comment by wxtim** [comment]
@stevewardle  can you resolve this conflict so that we can close this PR?

**Comment by stevewardle** [comment]
Ah okay, my mistake - I notice the branch against master is missing a fix which does exist on the other PR (which will allow the testing to work correctly).  Will copy said fix across now

**Comment by wxtim** [comment]
@matthewrmshin - needs a final review

**Pull request merged**